src/goto-instrument/contracts/dynamic-frames/dfcc_instrument_loop.cpp

```diff
@@ -0,0 +1,703 @@
+    language_mode);
+
+  add_exit_instructions(
+    loop_id,
+    cbmc_loop_id,
+    goto_function,
+    head,
+    loop.write_set_var,
+    loop.addr_of_write_set_var,
+    history_var_map,
+    entered_loop,
+    initial_invariant,
+    in_base_case,
+    old_decreases_vars,
+    new_decreases_vars);
+
+  goto_function.body.update();
+}
+
+std::map<exprt, exprt> dfcc_instrument_loopt::add_prehead_instructions(
+  const std::size_t loop_id,
+  goto_functionst::goto_functiont &goto_function,
+  symbol_tablet &symbol_table,
+  goto_programt::targett loop_head,
+  goto_programt::targett loop_latch,
+  goto_programt &assigns_instrs,
+  exprt &invariant,
+  const exprt::operandst &assigns,
+  const symbol_exprt &loop_write_set,
+  const symbol_exprt &addr_of_loop_write_set,
+  const symbol_exprt &entered_loop,
+  const symbol_exprt &initial_invariant,
+  const symbol_exprt &in_base_case,
+  const irep_idt &language_mode)
+{
+  auto loop_head_location(loop_head->source_location());
+  dfcc_remove_loop_tags(loop_head_location);
+
+  // ```
+  //   ... preamble ...
+  //
+  //   // Prehead block: Declare & initialize instrumentation variables
+  //   snapshot loop_entry history vars;
+  //   entered_loop = false
+  //   initial_invariant = loop_invariant;
+  //   in_base_case = true;
+  //   __ws_loop;
+  //   ws_loop := address_of(__ws_loop);
+  //   __write_set_create(ws_loop);
+  //   __write_set_add(ws_loop, loop_assigns);
+  //   __write_set_add(ws_loop, local_statics);
+  //   GOTO HEAD;
+  // ```
+
+  goto_programt pre_loop_head_instrs;
+  std::map<exprt, exprt> history_var_map;
+
+  // initialize loop_entry history vars;
+  {
+    // Process "loop_entry" history variables.
+    // We find and replace all "__CPROVER_loop_entry" in invariant.
+    replace_history_parameter(
+      symbol_table,
+      invariant,
+      history_var_map,
+      loop_head_location,
+      language_mode,
+      pre_loop_head_instrs,
+      ID_loop_entry);
+  }
+
+  // entered_loop = false
+  {
+    pre_loop_head_instrs.add(
+      goto_programt::make_decl(entered_loop, loop_head_location));
+    pre_loop_head_instrs.add(
+      goto_programt::make_assignment(entered_loop, false_exprt{}));
+  }
+
+  // initial_invariant = <loop_invariant>;
+  {
+    goto_convertt converter(symbol_table, log.get_message_handler());
+    // Create a snapshot of the invariant so that we can check the base case,
+    // if the loop is not vacuous and must be abstracted with contracts.
+    pre_loop_head_instrs.add(
+      goto_programt::make_decl(initial_invariant, loop_head_location));
+
+    // Although the invariant at this point will not have side effects,
+    // it is still a C expression, and needs to be "goto_convert"ed.
+    // Note that this conversion may emit many GOTO instructions.
+    code_assignt initial_invariant_assignment{initial_invariant, invariant};
+    initial_invariant_assignment.add_source_location() = loop_head_location;
+    converter.goto_convert(
```

Comment:
1) Nit pick: please move the `converter` declaration even closer to its use.
2) Please use `code_frontend_assignt(exprt lhs, exprt rhs, source_locationt loc)` instead of `code_assignt`: it has that additional source location (so you won't need `.add_source_location`), and you are really creating a pre-convert assignment here.